Fix continuation of JSON deserialization

devsko opened the pull request:
Fixes #42070 

When deserialization continues with a new chunk of data in the middle of a null token, a `JsonException` is thrown.
This PR attempts to fix this issue.

I verified that the change fixes the [original repro](https://github.com/frankbuckley/net5-json-repro).
I also verified that the two test cases (but with payload just beneath 2GB) mentioned by @am11 in this [comment](https://github.com/dotnet/runtime/issues/42070#issuecomment-691087902) work as expected with the changes in this PR.

The risk of regressions is low. Only continuation situations at null tokens are affected that throw or result in unexpected null values without the change.

devsko committed:
Repro #42070

devsko committed:
formatting

devsko committed:
namespace

devsko committed:
Fix

devsko committed:
never forget the header

layomia commented on src/libraries/System.Text.Json/tests/Serialization/ContinuationTests.cs:
```diff
@@ -0,0 +1,60 @@
+        private class TestClass
+        {
+            public string S { get; set; }
+            public string P { get; set; }
+            public TestClass C { get; set; }
```
Can we add tests for when P & C are value types (one primitive, one complex/POCO), and also when they are nullable value types?

Also, it would be good to add tests for the above scenarios when `TestClass` is deserialized with a parameterized ctor. This should help inform whether any change is needed in [ReadJsonAsObject](https://github.com/dotnet/runtime/blob/7861772d39d134c4517dfa76ab92109b899de833/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs#L268).

```cs
private class TestClassWithParameterizedCtor
{
    public string S { get; }
    public string P { get; }
    public TestClassWithParameterizedCtor C { get; }

    public TestClassWithParameterizedCtor(string s, string p, TestClassWithParameterizedCtor c) =>
        (S, P, C) = (s, p, c);
}
```

devsko replied:
I added a lot more tests and hope they cover most of the relevant paths. Especially all tests are now testing continuations at every position of the payload and not only at the null token. `ReadJsonAsObject` seems to work as it is.

layomia commented on src/libraries/System.Text.Json/tests/Serialization/ContinuationTests.cs:
```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
nit: let's call this `StreamTests.cs` to handle roundtrip stream cases since we already have `Stream.*Tests.cs`.

ahsonkhan replied:
Since we have other code paths to allow "partial progress"/continuation via the Utf8JsonReader overload, something a bit more generic as the filename that covers more uses, might actually be better than using `StreamTests`. 

Something like `ContinuationTests` seems reasonable, but **only** if it makes sense to add test cases for other overloads to this file too. See https://github.com/dotnet/runtime/pull/42158#discussion_r488907272

layomia commented on src/libraries/System.Text.Json/tests/Serialization/ContinuationTests.cs:
```diff
@@ -0,0 +1,60 @@
+    public static class ContinuationTests
+    {
+        [Fact]
+        public static async Task ContinationAtNullShouldWork()
```
```suggestion
        public static async Task ContinuationAtNullShouldWork()
```

layomia reviewed (commented):
Thanks for the PR! I requested a few more tests.

devsko committed:
More tests
- Test continuation at every position inside the tested object
- Many member with primitive and nullable types
- One more level of nested object
- All combinations of class/struct for tested and nested object
- tested and nested object with parametrized ctor for some properties

layomia commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,355 @@
+        // {"S":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","C":{"A":null,"B":"Hello","C":42,"D":null,"E":3.14E+17,"F":null,"G":true,"H":null,"I":[42,17],"J":{"A":null,"B":7}}}
+        // |<------------------------------------------------128 byte buffer------------------------------------------------------------->|
+        // {"S":"xxxxx","C":{"A":"Hello","B":null,"C":42,"D":null,"E":3.14E+17,"F":null,"G":true,"H":null,"I":[42,17],"J":{"A":null,"B":7}}}
+        public static IEnumerable<object[]> ContinuationSpacerLengths
+            => Enumerable.Range(5, 116 - 5 + 1).Select(length => new object[] { length });
```
How much longer does it take to run all the project tests in both `Debug`/`Release` modes as a result of using this value? If it's a lot, can we reduce the number of lengths returned here without losing the test coverage?

devsko replied:
After doubling all new tests because of `IgnoreNullValues=true|false` these are the local run results for target `net5.0`
||base|with continuation|
|:-:|:-:|:-:|
|**Debug**|`Total: 8706, Errors: 0, Failed: 0, Skipped: 0, Time: 70,223s`|`Total: 9826, Errors: 0, Failed: 0, Skipped: 0, Time: 72,229s`|
|**Release**|`Total: 8706, Errors: 0, Failed: 0, Skipped: 0, Time: 28,578s`|`Total: 9826, Errors: 0, Failed: 0, Skipped: 0, Time: 31,961s`|

I would say 2-3 seconds is good enough. How would you decrease the tested positions without knowing what could go wrong in advance?
BTW I cannot run the tests for target `net461` from 6.0 branch locally because I get tons of errors. I "hope" it's just a glitch with my machine
```
System.IO.FileLoadException : Die Datei oder Assembly "System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" oder eine Abhängigkeit davon wurde nicht gefunden. Fehler bei der Überprüfung des starken Namens. (Ausnahme von HRESULT: 0x8013141A)
```

layomia replied:
I think 2-3s is worth the coverage here.

---

> cannot run the tests for target net461

Did you try building the libraries from the repo root before running the tests?

`build -s clr+libs -rc Release -f net461`

devsko replied:
`-s` is for signing? I'll give it a try. Can't build clr at all after upgrade to VS 16.8 preview 3

devsko replied:
Okay `-s` is `-subset` but optional. I rebuilt everything and still cannot run tests for net461. I would say it fails since #41616 got merged. The binding redirect is generated in S.T.J.Tests.dll.config but has no effect during the test run in VS nor with `dotnet build /t:test`.

devsko replied:
Opened #42250 but still don't know if it's only my environment.

layomia replied:
Thanks for filing the issue. It looks like there was indeed an issue which is being looked into.

layomia commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,355 @@
+                var readOptions = new JsonSerializerOptions
+                {
+                    DefaultBufferSize = 128,
+                    IgnoreNullValues = true
```
Let's also test when `IgnoreNullValues` is false.

steveharter commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:
```diff
@@ -234,7 +234,7 @@ public override bool ReadJsonAndSetMember(object obj, ref ReadStack state, ref U
             {
                 success = true;
-                if (!isNullToken || !IgnoreDefaultValuesOnRead || !Converter.CanBeNull)
+                if (!isNullToken || !IgnoreDefaultValuesOnRead || !Converter.CanBeNull || state.IsContinuation)
```
In the [`if` statement above here](https://github.com/dotnet/runtime/blob/4b3e724919da9eacc1e63a7bbb9c5b4b08b10acd/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs#L223-L233)  please add an Assert:
```cs
            else if (Converter.CanUseDirectReadOrWrite && state.Current.NumberHandling == null)
            {
                Debug.Assert(!state.IsContinuation);
```

The Assert will be true since `CanUseDirectReadOrWrite` is false when using a Stream.

devsko replied:
Added assertions here and in `ReadJsonAsObject`.

steveharter commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,355 @@
+    public static partial class StreamTests
+    {
+        // To hit all possible continuation positions inside the tested object,
+        // the outer-class spacer needs to be between 5 and 116 bytes long.
```
Thanks -- doing padding like this makes sense and was previously discussed, but never implemented AFAIK.

layomia replied:
We added some to [test the parameterized ctors feature](https://github.com/dotnet/runtime/pull/33444/files#diff-2f4d68bf74486907056cc46853c29e89R150), but it would be good to extend these tests as suggested in https://github.com/dotnet/runtime/pull/42158#discussion_r488907272.

steveharter commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,355 @@
+        // |<------------------------------------------------128 byte buffer------------------------------------------------------------->|
+        // {"S":"xxxxx","C":{"A":"Hello","B":null,"C":42,"D":null,"E":3.14E+17,"F":null,"G":true,"H":null,"I":[42,17],"J":{"A":null,"B":7}}}
+        public static IEnumerable<object[]> ContinuationSpacerLengths
+            => Enumerable.Range(5, 116 - 5 + 1).Select(length => new object[] { length });
```
Please add some comments or constants on the 116 and 5 literals.

Once we get the first continuation, the buffer size will double to 256. With a small "S" string of 5 "xxxxx", length appears 128, so:
- To get continuation on the last EndObject tokens, don't we need 256 extra padding bytes?
- To get continuation on the first StartObject, don't we at least 1 extra padding byte?

devsko replied:
> Once we get the first continuation, the buffer size will double to 256.

Right now there is always only one continuation per test case. The padding determines where it happens.

- The 5 bytes padding leads to a first chunk that contains anything up to `...,"J":{"A":null,"B":7}}`. After continuation there is only the very last `EndObject` left.
- The 116 bytes padding leads to a fist chunk that ends with `...xxxxx","C":`. The continuation starts then with the `StartObject` of the tested object.

Maybe I don't get what you mean.

steveharter replied:
OK thanks. I see now that there is only one continuation.

Jozkee commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,355 @@
+                    }
+                };
+                await JsonSerializer.SerializeAsync(stream, obj);
+                var json = Encoding.UTF8.GetString(stream.ToArray()); 
```
`json` is never used.

Jozkee commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,355 @@
+                    IgnoreNullValues = true
+                };
+
+                var obj = await JsonSerializer.DeserializeAsync<Outer<TestClass<NestedClass>>>(stream, readOptions);
```
nit:
```suggestion
                Outer<TestClass<NestedClass>> obj = await JsonSerializer.DeserializeAsync<Outer<TestClass<NestedClass>>>(stream, readOptions);
```

devsko committed:
Addressed feedback

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,368 @@
+        // |<------------------------------------------------128 byte buffer------------------------------------------------------------->|
+        // {"S":"xxxxx","C":{"A":"Hello","B":null,"C":42,"D":null,"E":3.14E+17,"F":null,"G":true,"H":null,"I":[42,17],"J":{"A":null,"B":7}}}
+
+        private const int MinPaddingLength = 5;
+        private const int MaxPaddingLength = 116;
+
+        private static IEnumerable<int> ContinuationPaddingLengths
+            => Enumerable.Range(MinPaddingLength, MaxPaddingLength - MinPaddingLength + 1);
+
+        private static IEnumerable<bool> IgnoreNullValues
+            => new[] { true, false };
```
These are great tests and a clever way to get a lot more continuation coverage! Nicely done.

If you are up for it, I think it would be really valuable to do a similar thing with the `Utf8JsonReader` overload of `JsonSerialier.Deserialize` (possibly in a separate PR), which also supports continuation/partial buffers via the `ReadOnlySequence<byte>` (where the whole JSON is available in memory, but split in non-contiguous chunks).
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize?view=netcore-3.1#System_Text_Json_JsonSerializer_Deserialize_System_Text_Json_Utf8JsonReader__System_Type_System_Text_Json_JsonSerializerOptions_
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader.-ctor?view=netcore-3.1#System_Text_Json_Utf8JsonReader__ctor_System_Buffers_ReadOnlySequence_System_Byte__System_Boolean_System_Text_Json_JsonReaderState_

devsko replied:
Thanks!
I will have a look.

ahsonkhan commented:
> I verified that the change fixes the original repro.
I also verified that the two test cases (but with payload just beneath 2GB) mentioned by @am11 in this comment work as expected with the changes in this PR.

If feasible (and quick to run), can we also add to the test test a standalone copy of the "regression" test that was initially reported on the issue: https://github.com/dotnet/runtime/issues/42070#issuecomment-691147167

That way, that exact customer reported scenario is guarded against in the test.

Similar to these:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.ObjectModel.Read.cs#L19
https://github.com/dotnet/runtime/blob/5285c141fe2b2f48e7092588a2ef5dcf5cc9c60a/src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.TryGet.Date.cs#L150-L152
https://github.com/dotnet/runtime/blob/5285c141fe2b2f48e7092588a2ef5dcf5cc9c60a/src/libraries/System.Text.Json/tests/Utf8JsonWriterTests.cs#L3091-L3097
https://github.com/dotnet/runtime/blob/5285c141fe2b2f48e7092588a2ef5dcf5cc9c60a/src/libraries/System.Text.Json/tests/Serialization/Object.ReadTests.cs#L111-L112
https://github.com/dotnet/runtime/blob/5285c141fe2b2f48e7092588a2ef5dcf5cc9c60a/src/libraries/System.Text.Json/tests/Serialization/ReadValueTests.cs#L671-L673

devsko commented:
>can we also add to the test test a standalone copy of the "regression" test that was initially reported 

I was thinking about that but there are some obstacles. The original repro depends on some third party packages. Besides it is not deterministic because nobody knew the root cause at this time. Sometimes it runs for minutes until randomly a null token gets hit by the continuation.

frankbuckley commented:
> I was thinking about that but there are some obstacles. 

@devsko / @ahsonkhan - thank you for looking into this.

I have added code to save failing examples to [the repro](https://github.com/frankbuckley/net5-json-repro) - the [failing json examples are around 100KB](https://github.com/frankbuckley/net5-json-repro/tree/master/examples/2020-09-16-104226). Maybe a few of these could be used for regression tests?

Also, the captured output includes some examples of a NullReferenceException - in addition to examples of the more common InvalidOperationException - [for example, 0228](https://github.com/frankbuckley/net5-json-repro/blob/29d116cb2b1d8e14b2aae31515d6753a2e566378/examples/2020-09-16-104226/2020-09-16-104226.txt#L402).

@am11 [noted a NullReferenceException](https://github.com/dotnet/runtime/issues/42070#issuecomment-691087902) and I have seen this twice with real data.

Does the fix for the InvalidOperationExceptions also fix the NullReferenceExceptions?

devsko commented:
>I have added code to save failing examples to the repro - the failing json examples are around 100KB. Maybe a few of these could be used for regression tests?

Great! Thank you. I will add a test for each observed exception type.
>Does the fix for the InvalidOperationExceptions also fix the NullReferenceExceptions?

It does.

devsko committed:
Test with original repro data from #42070

devsko committed:
custom converter to ensure the padding is written in front of the tested object

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/StreamTests.cs:
```diff
@@ -0,0 +1,511 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Follow up to https://github.com/dotnet/runtime/pull/42158#discussion_r487553719 (to raise visibility, since that is resolved)

@layomia, @devsko - what do you think?

> Since we have other code paths to allow "partial progress"/continuation via the `Utf8JsonReader` overload, something a bit more generic as the filename that covers more uses, might actually be better than using `StreamTests`. 

> Something like `ContinuationTests` seems reasonable, but **only** if it makes sense to add test cases for other overloads to this file too. See https://github.com/dotnet/runtime/pull/42158#discussion_r488907272

devsko replied:
I already started to work on a follow up PR for the other continuations. I'm happy to change the name then. Hope I get it done this week because I'm away some time starting next week 😁

layomia replied:
`ContinuationTests` is fine with me.

devsko replied:
Do you want to merge this PR into 5?

layomia replied:
Sorry I didn't see your above response. Yes we plan to port this PR to .NET 5. If Ship-room gives the OK, we should be able to port it this week. I have some additional feedback that will need another commit, so I'd appreciate doing the rename now.

> Hope I get it done this week because I'm away some time starting next week 😁

Nice, enjoy!

devsko replied:
> I have some additional feedback that will need another commit, so I'd appreciate doing the rename now.

Okay I wait for your feedback and will address it tomorrow (it's 8pm here)

layomia replied:
https://github.com/dotnet/runtime/pull/42158#discussion_r489633879

layomia commented on src/libraries/System.Text.Json/tests/Resources/Issue42070_0004.json:
We store other test payloads in this [`Strings.resx` file](https://raw.githubusercontent.com/dotnet/runtime/a7b9c36d19bfddef73339e7a59f03e27bdd7b1f6/src/libraries/System.Text.Json/tests/Resources/Strings.resx). Based on some other entries, the data name should be `SomeDescriptivePayloadSize` e.g. `BigPayloadWithNulls101KB`.

The usage would be `string json = SR.BigPayloadWithNulls101KB;`

devsko committed:
rename

devsko committed:
test data moved to Strings.resx

layomia commented on src/libraries/System.Text.Json/tests/Serialization/ContinuationTests.cs:
```diff
@@ -0,0 +1,511 @@
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static class ContinuationTests
```
The first four tests in this file are very similar, the only difference being the inner type passed to outer, and also an `Assert.Null` test if the innermost type is a reference type. To avoid the duplication: can we consolidate them into one generic test helper, and run the test like this?

```cs
[Theory]
[MemberData(nameof(TestData))]
public static async Task ContinuationShouldWorkAtAnyPosition_ParameterlessCtor(int paddingLength, bool ignoreNullValues)
{
    await RunTest<TestClass<NestedClass>, NestedClass>(paddingLength, ignoreNullValues);
    await RunTest<TestClass<NestedValueType>, NestedValueType>(paddingLength, ignoreNullValues);
    await RunTest<TestValueType<NestedClass>, NestedClass>(paddingLength, ignoreNullValues);
    await RunTest<TestValueType<NestedValueType>, NestedValueType>(paddingLength, ignoreNullValues);
}

private static async Task RunTest<TInner, TInnermost>(int paddingLength, bool ignoreNullValues)
    where TInner : ITestInterface<TInnermost>, new()
    where TInnermost : INestedInterface, new()
{
    var stream = new MemoryStream();
    {
        var obj = new Outer<TInner>
        {
            S = new string('x', paddingLength),
            C = new()
            {
                A = "Hello",
                B = null,
                C = 42,
                D = null,
                E = 3.14e+17f,
                F = null,
                G = true,
                H = null,
                I = new int[] { 42, 17 },
                J = new()
                {
                    A = null,
                    B = 7,
                }
            }
        };
        await JsonSerializer.SerializeAsync(stream, obj, new JsonSerializerOptions { Converters = { new OuterConverter<TestClass<NestedClass>>() } });
    }

    stream.Position = 0;
    {
        var readOptions = new JsonSerializerOptions
        {
            DefaultBufferSize = 128,
            IgnoreNullValues = ignoreNullValues,
        };

        Outer<TestClass<NestedClass>> obj = await JsonSerializer.DeserializeAsync<Outer<TestClass<NestedClass>>>(stream, readOptions);

        Assert.Equal(new string('x', paddingLength), obj.S);
        Assert.Equal("Hello", obj.C.A);
        Assert.Null(obj.C.B);
        Assert.Equal(42, obj.C.C);
        Assert.Null(obj.C.D);
        Assert.Equal(3.14e17f, obj.C.E);
        Assert.Null(obj.C.F);
        Assert.True(obj.C.G);
        Assert.Null(obj.C.H);
        Assert.Collection(obj.C.I, v => Assert.Equal(42, v), v => Assert.Equal(17, v));

        bool innerIsRefType = typeof(TInner).IsValueType == false;
        if (innerIsRefType)
        {
            Assert.NotNull(obj.C.J);
        }

        Assert.Null(obj.C.J.A);
        Assert.Equal(7, obj.C.J.B);
    }
}

private interface ITestInterface<TNested>
{
    public string A { get; set; }
    public string B { get; set; }
    public int C { get; set; }
    public int? D { get; set; }
    public float E { get; set; }
    public float? F { get; set; }
    public bool G { get; set; }
    public bool? H { get; set; }
    public int[] I { get; set; }
    public TNested J { get; set; }
}

private class TestClass<TNested> : ITestInterface<TNested>
{
    public string A { get; set; }
    public string B { get; set; }
    public int C { get; set; }
    public int? D { get; set; }
    public float E { get; set; }
    public float? F { get; set; }
    public bool G { get; set; }
    public bool? H { get; set; }
    public int[] I { get; set; }
    public TNested J { get; set; }
}

private class TestValueType<TNested> : ITestInterface<TNested>
{
    public string A { get; set; }
    public string B { get; set; }
    public int C { get; set; }
    public int? D { get; set; }
    public float E { get; set; }
    public float? F { get; set; }
    public bool G { get; set; }
    public bool? H { get; set; }
    public int[] I { get; set; }
    public TNested J { get; set; }
}

private interface INestedInterface
{
    public string A { get; set; }
    public int B { get; set; }
}

private class NestedClass : INestedInterface
{
    public string A { get; set; }
    public int B { get; set; }
}

private struct NestedValueType : INestedInterface
{
    public string A { get; set; }
    public int B { get; set; }
}
```

I didn't look further, but there might be a way to abstract further and reduce duplicate LOC for the parameterized ctor test; up to you.

devsko replied:
>The first four tests in this file are very similar

I already completely refactored this for the same reason. Follow up PR is coming soon.

layomia reviewed (commented):
A couple test comments, otherwise LGTM.

layomia reviewed (approved):
Above feedback on tests can be addressed in follow up. Merging now to get this PR ready for ship-room consideration for .NET 5.

Thanks @devsko

layomia commented:
> I already started to work on a follow up PR for the other continuations.

@devsko  - can you include a test for when partial buffer ends on a valid prefix for `null`, but turns out to be invalid JSON on continuation? This is likely going to be `JsonException`, but it would be good to encode the behavior in a test.

e.g. 

Initial buffer: `{"Foo":"Bar","Baz":nu`

Next buffer: `lllt,"Qux":"Quux"}`
